File: Sorting/quicksort.cpp
#include<iostream>
using namespace std;

void printarray(int arr[] , int n)
{
    for (int i = 0; i < n; i++)
    {
        cout<<arr[i]<<" ";
    }
    
}

int partition(int arr[] ,  int low , int high)
{
    int pivot = arr[low];
    int i = low+1;
    int j = high;
    int temp;
    
    do
    {
        while (arr[j] > pivot)
        {
            j--;
        }

        while (arr[i] <= pivot)
        {
            i++;
        }
        
        if (i<j)
        {
            temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }

    } while (i < j);

    temp = arr[low];
    arr[low] = arr[j];
    arr[j] = temp;
    return j;

}

void quicksort(int arr[] , int low , int high)
{
    int partitionidx;

    if (low<high)
    {
        partitionidx = partition(arr , low , high);
        quicksort(arr , low , partitionidx-1);
        quicksort(arr , partitionidx+1 , high);
    }
    
}

int main()
{
    int arr[10];
    int n = 10;

    for (int i = 0; i < n; i++)
    {
        cout<<"Enter an element ";
        cin>>arr[i];
    }
    
    quicksort(arr , 0 , 9);
    printarray(arr , n);
    return 0;
}

File: Sorting/QuickSort.java
package com.gecg;

public class QuickSort {

    public static int partition(int arr[] , int low , int high)
    {
        int pivot = arr[high];
        int i=low-1;

        for (int j = low; j < high ; j++) {
            if (arr[j] < pivot)
            {
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        i++;
        int temp = arr[i];
        arr[i] = pivot;
        arr[high] = temp;

        return i;
    }
    public static void Quicksort(int arr[] , int low , int high)
    {
        if (low < high)
        {
            int pivot_idx = partition(arr , low , high);
            Quicksort(arr , low , pivot_idx-1);
            Quicksort(arr , pivot_idx+1 , high);
        }
    }

    public static void printArray(int arr[])
    {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
    public static void main(String[] args) {
        int arr[] = {5,93,45,2,1,264};
        printArray(arr);
        Quicksort(arr , 0 , arr.length-1);
        printArray(arr);

    }
}
